arbos/tx_processor.go

```diff
@@ -269,10 +269,29 @@ func (p *TxProcessor) StartTxHook() (endTxNow bool, gasUsed uint64, err error, r
 
 		// pay for the retryable's gas and update the pools
 		gascost := arbmath.BigMulByUint(basefee, usergas)
-		if err := transfer(&tx.From, &networkFeeAccount, gascost); err != nil {
-			// should be impossible because we just checked the tx.From balance
-			glog.Error("failed to transfer gas cost to network fee account", "err", err)
-			return true, 0, nil, ticketId.Bytes()
+		networkCost := gascost
+		if p.state.ArbOSVersion() >= 11 {
+			infraFeeAccount, _ := p.state.InfraFeeAccount()
+			if infraFeeAccount != (common.Address{}) {
+				infraFee, err := p.state.L2PricingState().MinBaseFeeWei()
+				p.state.Restrict(err)
+				if arbmath.BigLessThan(basefee, infraFee) {
+					infraFee = basefee
+				}
+				infraCost := arbmath.BigMulByUint(infraFee, usergas)
+				if err := transfer(&tx.From, &infraFeeAccount, infraCost); err != nil {
+					glog.Error("failed to transfer gas cost to infrastructure fee account", "err", err)
+					return true, 0, nil, ticketId.Bytes()
+				}
+				networkCost = arbmath.BigSub(networkCost, infraCost)
```

Comment:
Could we use the takeFunds function here, just in case infraCost is somehow greater than networkCost?

Reply:
makes sense, fixed, thanks :)